feat(metro-config, cli): CSS support for Metro web

EvanBacon opened the pull request:
# Why

This PR moves the CSS support from Expo Router over to `expo/metro-config` behind a beta flag. This is because we need deeper integration with Expo CLI and Metro in order to emit static CSS files in production bundles. CSS is required for media queries -> rehydration.

- Related: https://github.com/expo/router/pull/397 https://github.com/expo/router/pull/223

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

# How

Instead of a babel transformer, we now use a custom "transformer" which gives us the ability to add extra metadata to a Metro module on export. This means we can pass the raw and processed CSS for writing to disk.

The development version of CSS still uses script injection via JavaScript, meaning static rendering cannot currently be tested in development.

CSS Modules are implemented (web-only currently) using lightningcss. The export is generated to work with React Native for web. Consider the following block:

```js
export default { ...StyleSheet.create({ container: { $$css: true, _: 'hashed-container-id' }  }) }
```

CSS Variables are not currently hashed, enabling the user to define variables in a global CSS file and access them in CSS Modules (subject to change in order to support native). CSS variables are accessible from the default export as strings: `styles['--color'] === '--color'`.


## Docs

I've chosen to document in the versioned metro doc instead of the unversioned guide, this seems like a reasonable spot given the highly experimental nature of this feature.


<!--
How did you build this feature or fix this bug and why?
-->

# Test Plan

- Unit tests for conversion.
- [x] Test in/out for transformer.
- [ ] Test static export.
- TBD for E2E


<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

byCedric commented on packages/@expo/cli/e2e/fixtures/with-router/styles/global.css:
```diff
@@ -0,0 +1,3 @@
+.container {
+    background-color: aqua;
```
![Uploading image.png…]()

motiz88 commented on packages/@expo/metro-config/src/transform-worker/transform-worker.ts:
```diff
@@ -0,0 +1,219 @@
+  getCacheKey(config: JsTransformerConfig): string {
+    const { babelTransformerPath, minifierPath, ...remainingConfig } = config;
+
+    const filesKey = getCacheKey([
+      require.resolve(babelTransformerPath),
+      require.resolve(minifierPath),
+      require.resolve('metro-transform-worker/src/utils/getMinifier'),
+      require.resolve('metro-transform-worker/src/utils/assetTransformer'),
+      require.resolve('metro/src/ModuleGraph/worker/generateImportNames'),
+      require.resolve('metro/src/ModuleGraph/worker/JsFileWrapping'),
+      ...metroTransformPlugins.getTransformPluginCacheKeyFiles(),
+    ]);
+
+    const babelTransformer = require(babelTransformerPath);
+    return [
+      filesKey,
+      stableHash(remainingConfig).toString('hex'),
+      babelTransformer.getCacheKey ? babelTransformer.getCacheKey() : '',
+    ].join('$');
```
You should replace most of this with a call to the built-in transformer's `getCacheKey` function instead of duplicating its behaviour.

EvanBacon commented on packages/@expo/metro-config/src/transform-worker/transform-worker.ts:
```diff
@@ -0,0 +1,219 @@
+          },
+        },
+        // In production, we change the type so it's not included in the JS bundle.
+        //   type: 'js/module/css',
```
```suggestion
```

EvanBacon committed:
css support

update tests

Update webTemplate-test.ts

update tests

fix some tests

build files

Update CHANGELOG.md

EvanBacon committed:
fix default

EvanBacon committed:
Update _layout.js

EvanBacon committed:
Update yarn.lock

EvanBacon committed:
rebuild prebuild

EvanBacon committed:
Merge branch 'main' into @evanbacon/metro-config/css-transform-worker

EvanBacon committed:
Update withAndroidIcons.js.map
